proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,457 @@
+have kind ``TYPE 'UnliftedRep``. Indeed, for types of this kind GHC will
+already omit the zero tag check today! So all that is left is to extend GHC in
+a way that
+
+::
+  
+  data unlifted IntSet
+    = Branch IntSet !Int IntSet
+    | Leaf
+
+  member :: Int -> IntSet -> Bool
+  member !k Leaf = False
+  member k (Branch l k' r)
+    | k == k'   = True
+    | k <  k'   = member k l
+    | otherwise = member k r
+
+does the Right Thing, namely disallowing unevaluated boxes to be passed around
+by using call-by-value. This way we can have the guarantee that all unlifted
+boxes are always properly tagged.
+
+Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
+suggest that the unlifted variant is 5-10% faster, just due to omission of the
+zero tag check.
+
+Proposed Change Specification
+-----------------------------
+
+Henceforth, data type declaration refers to both data type and data family instance declarations.
+
+Syntax
+~~~~~~
+
+The language extension ``-XUnliftedDatatypes`` introduces a new contextual
+keyword ``unlifted``, only to be used in data type declarations. Revised
+grammar rules:
+
+::
+
+ topdecl   -> 'data' [ 'lifted' | 'unlifted' ] [ context => ] ...
+ topdecl   -> 'data' 'instance' [ 'lifted' | 'unlifted' ] [ context => ] ...
+ decl_inst -> 'data' [ 'instance' ] [ 'lifted' | 'unlifted' ] [ context => ] ...
+
+GADT-style declarations can optionally specify a kind signature.
+
+The ``lifted`` keyword can be dropped after parsing; it's just for symmetry.
+
+Note that (like data families with ``instance``) we are potentially stealing
+syntax: ``data unlifted = T Int`` could currently parse ``unlifted`` as the
+data type to be defined and bail out with a parser validation error
+(``Malformed head of type or class declaration: unlifted``). Under this
+proposal, it would turn into a parser error (``parse error on input ‘=’``).
+
+(I couldn't come up with an example where we actually turn a previously
+semantically correct program into one with a parser error.)
+
+Static semantics
+~~~~~~~~~~~~~~~~
+
+Name resolution can ignore the ``unlifted`` keyword.
+
+Similar to ``-XUnliftedNewtypes``, the return kind of a data type declaration's
+kind signature (which may be given explicitly by the user or be inferred) is
+``TYPE 'UnliftedRep`` when there was a leading ``unlifted`` keyword.
+
+Data family instances may be declared unlifted, in which case the data family
+application's result kind must reduce to ``TYPE 'UnliftedRep``. See
+`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
+and ``Note [Implementation of UnliftedNewtypes]`` for details involving
+type-checking the parent data family.
+
+The static semantics of other types of unlifted kind, such as the inability to
+delare them at the top-level, apply. The top-level restriction is not
```

Comment:
```suggestion
declare them at the top-level, apply. The top-level restriction is not
```

Reply:
Fixed in https://github.com/ghc-proposals/ghc-proposals/pull/265/commits/356566d78fb08e6703c1c4acb808ddfdf1119efa.